File: base-v3/src/main/java/com/adyen/checkout/base/model/payments/response/RedirectAction.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 4/6/2019.
 */

package com.adyen.checkout.base.model.payments.response;

import android.os.Parcel;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.base.util.ActionTypes;
import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.model.JsonUtils;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings({"MemberName", "PMD.DataClass"})
public class RedirectAction extends Action {
    @NonNull
    public static final Creator<RedirectAction> CREATOR = new Creator<>(RedirectAction.class);

    public static final String ACTION_TYPE = ActionTypes.REDIRECT;

    private static final String METHOD = "method";
    private static final String URL = "url";

    @NonNull
    public static final Serializer<RedirectAction> SERIALIZER = new Serializer<RedirectAction>() {

        @NonNull
        @Override
        public JSONObject serialize(@NonNull RedirectAction modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                // Get parameters from parent class
                jsonObject.putOpt(Action.TYPE, modelObject.getType());
                jsonObject.putOpt(Action.PAYMENT_DATA, modelObject.getPaymentData());
                jsonObject.putOpt(Action.PAYMENT_METHOD_TYPE, modelObject.getPaymentMethodType());

                jsonObject.putOpt(METHOD, modelObject.getMethod());
                jsonObject.putOpt(URL, modelObject.getUrl());
            } catch (JSONException e) {
                throw new ModelSerializationException(RedirectAction.class, e);
            }
            return jsonObject;
        }

        @NonNull
        @Override
        public RedirectAction deserialize(@NonNull JSONObject jsonObject) {
            final RedirectAction redirectAction = new RedirectAction();

            // getting parameters from parent class
            redirectAction.setType(jsonObject.optString(Action.TYPE, null));
            redirectAction.setPaymentData(jsonObject.optString(Action.PAYMENT_DATA, null));
            redirectAction.setPaymentMethodType(jsonObject.optString(Action.PAYMENT_METHOD_TYPE, null));

            redirectAction.setMethod(jsonObject.optString(METHOD, null));
            redirectAction.setUrl(jsonObject.optString(URL, null));
            return redirectAction;
        }
    };

    private String method;
    private String url;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public String getMethod() {
        return method;
    }

    @Nullable
    public String getUrl() {
        return url;
    }

    public void setMethod(@Nullable String method) {
        this.method = method;
    }

    public void setUrl(@Nullable String url) {
        this.url = url;
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/model/payments/response/RedirectAction.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 4/6/2019.
 */
package com.adyen.checkout.components.model.payments.response

import android.os.Parcel
import com.adyen.checkout.components.util.ActionTypes
import com.adyen.checkout.core.exception.ModelSerializationException
import com.adyen.checkout.core.model.JsonUtils.writeToParcel
import com.adyen.checkout.core.model.getStringOrNull
import org.json.JSONException
import org.json.JSONObject

data class RedirectAction(
    override var type: String? = null,
    override var paymentData: String? = null,
    override var paymentMethodType: String? = null,
    var method: String? = null,
    var url: String? = null,
) : Action() {

    override fun writeToParcel(dest: Parcel, flags: Int) {
        writeToParcel(dest, SERIALIZER.serialize(this))
    }

    companion object {
        const val ACTION_TYPE = ActionTypes.REDIRECT
        private const val METHOD = "method"
        private const val URL = "url"

        @JvmField
        val CREATOR = Creator(RedirectAction::class.java)

        @JvmField
        val SERIALIZER: Serializer<RedirectAction> = object : Serializer<RedirectAction> {
            override fun serialize(modelObject: RedirectAction): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(TYPE, modelObject.type)
                        putOpt(PAYMENT_DATA, modelObject.paymentData)
                        putOpt(PAYMENT_METHOD_TYPE, modelObject.paymentMethodType)
                        putOpt(METHOD, modelObject.method)
                        putOpt(URL, modelObject.url)
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(RedirectAction::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): RedirectAction {
                return RedirectAction(
                    type = jsonObject.getStringOrNull(TYPE),
                    paymentData = jsonObject.getStringOrNull(PAYMENT_DATA),
                    paymentMethodType = jsonObject.getStringOrNull(PAYMENT_METHOD_TYPE),
                    method = jsonObject.getStringOrNull(METHOD),
                    url = jsonObject.getStringOrNull(URL),
                )
            }
        }
    }
}
